src/lib/unlang/xlat_builtin.c

```diff
@@ -1117,97 +1117,98 @@ static xlat_action_t xlat_func_explode(TALLOC_CTX *ctx, fr_dcursor_t *out,
+static xlat_action_t xlat_func_integer(TALLOC_CTX *ctx, fr_dcursor_t *out,
+				       request_t *request, UNUSED void const *xlat_inst,
+				       UNUSED void *xlat_thread_inst, fr_value_box_list_t *in)
 {
-	fr_pair_t	*vp;
+	fr_value_box_t	*in_vb = fr_dlist_head(in);
+	fr_value_box_t	*vb;
 
 	uint64_t	int64 = 0;	/* Needs to be initialised to zero */
-	uint32_t	int32 = 0;	/* Needs to be initialised to zero */
-
-	fr_skip_whitespace(fmt);
+	char		buff[40];	/* Temporary string representation special conversions */
 
-	if ((xlat_fmt_get_vp(&vp, request, fmt) < 0) || !vp) return 0;
+	switch (in_vb->type) {
+	case FR_TYPE_UINT8:
+	case FR_TYPE_UINT16:
+	case FR_TYPE_UINT32:
+	case FR_TYPE_UINT64:
+	case FR_TYPE_INT8:
+	case FR_TYPE_INT16:
+	case FR_TYPE_INT32:
+	case FR_TYPE_INT64:
+	case FR_TYPE_TIME_DELTA:
+	case FR_TYPE_SIZE:
+	case FR_TYPE_BOOL:
+		/*
+		 *	Ensure enumeration is NULL so that the integer
+		 *	version of a box is returned
+		 */
+		fr_value_box_cast_in_place(ctx, in_vb, in_vb->type, NULL);
+		break;
 
-	switch (vp->vp_type) {
 	case FR_TYPE_DATE:
 	case FR_TYPE_STRING:
-	{
-		fr_value_box_t vb;
-
-		if (fr_value_box_cast(NULL, &vb, FR_TYPE_UINT64, NULL, &vp->data) < 0) {
-			RPEDEBUG("Invalid input for printing as an integer");
-			return -1;
+		if (fr_value_box_cast_in_place(ctx, in_vb, FR_TYPE_UINT64, NULL) < 0) {
+		error:
+			RPEDEBUG("Invalid input for casting to an integer");
+			return XLAT_ACTION_FAIL;
 		}
-
-		return snprintf(*out, outlen, "%" PRIu64, vb.vb_uint64);
-	}
+		break;
 
 	case FR_TYPE_OCTETS:
-		if (vp->vp_length > 8) {
-			break;
-		}
+		if (in_vb->vb_length > 8) goto error;
 
-		if (vp->vp_length > 4) {
-			memcpy(&int64, vp->vp_octets, vp->vp_length);
-			return snprintf(*out, outlen, "%" PRIu64, htonll(int64));
+		if (in_vb->vb_length > 4) {
+			fr_value_box_cast_in_place(ctx, in_vb, FR_TYPE_INT64, NULL);
+			break;
 		}
 
-		memcpy(&int32, vp->vp_octets, vp->vp_length);
-		return snprintf(*out, outlen, "%i", htonl(int32));
-
-	case FR_TYPE_UINT64:
-		return snprintf(*out, outlen, "%" PRIu64, vp->vp_uint64);
+		fr_value_box_cast_in_place(ctx, in_vb, FR_TYPE_INT32, NULL);
+		break;
 
 	/*
 	 *	IP addresses are treated specially, as parsing functions assume the value
 	 *	is bigendian and will convert it for us.
 	 */
 	case FR_TYPE_IPV4_ADDR:
 	case FR_TYPE_IPV4_PREFIX:	/* Same addr field */
-		return snprintf(*out, outlen, "%u", ntohl(vp->vp_ipv4addr));
-
-	case FR_TYPE_UINT32:
-		return snprintf(*out, outlen, "%u", vp->vp_uint32);
-
-	case FR_TYPE_UINT8:
-		return snprintf(*out, outlen, "%u", (unsigned int) vp->vp_uint8);
-
-	case FR_TYPE_UINT16:
-		return snprintf(*out, outlen, "%u", (unsigned int) vp->vp_uint16);
+		snprintf(buff, 40, "%lu", (long unsigned int)ntohl(in_vb->vb_ip.addr.v4.s_addr));
+		goto output_buff;
 
 	/*
 	 *	Ethernet is weird... It's network related, so it
 	 *	should be bigendian.
 	 */
 	case FR_TYPE_ETHERNET:
-		int64 = vp->vp_ether[0];
+		int64 = in_vb->vb_ether[0];
```

Comment:
I think just cast to a UINT64 and ignore the comment.  If there's not that capability in the box casting function already then add it.